File: src/Scripting/RubyRB/factory_method.rb
class DatabaseCreator
  def create_database
    raise NotImplementedError
  end

  def use_database
    database = create_database
    database.connect
    database.query
  end
end

class PostgresCreator < DatabaseCreator
  def create_database
    Object.new.tap do |database|
      database.define_singleton_method(:connect) { puts "PostgreSQL connect" }
      database.define_singleton_method(:query) { puts "PostgreSQL query" }
    end
  end
end

class MySqlCreator < DatabaseCreator
  def create_database
    Object.new.tap do |database|
      database.define_singleton_method(:connect) { puts "MySQL connect" }
      database.define_singleton_method(:query) { puts "MySQL query" }
    end
  end
end

PostgresCreator.new.use_database
MySqlCreator.new.use_database

File: src/Scripting/PythonPY/factory_method.py
from abc import ABC, abstractmethod


class Database(ABC):
    @abstractmethod
    def connect(self) -> None:
        raise NotImplementedError

    @abstractmethod
    def query(self) -> None:
        raise NotImplementedError


class PostgresDatabase(Database):
    def connect(self) -> None:
        print("PostgreSQL connect")

    def query(self) -> None:
        print("PostgreSQL query")


class MySqlDatabase(Database):
    def connect(self) -> None:
        print("MySQL connect")

    def query(self) -> None:
        print("MySQL query")


class DatabaseCreator(ABC):
    @abstractmethod
    def create_database(self) -> Database:
        raise NotImplementedError

    def use_database(self) -> None:
        database = self.create_database()
        database.connect()
        database.query()


class PostgresCreator(DatabaseCreator):
    def create_database(self) -> Database:
        return PostgresDatabase()


class MySqlCreator(DatabaseCreator):
    def create_database(self) -> Database:
        return MySqlDatabase()


if __name__ == "__main__":
    PostgresCreator().use_database()
    MySqlCreator().use_database()
